test/server/cards/02_SHD/leaders/FinnThisIsARescue.spec.ts

```diff
@@ -0,0 +1,127 @@
+describe('Finn, This is a Rescue', function () {
+    integration(function (contextRef) {
+        describe('Finn\'s undeployed ability', function () {
+            beforeEach(function () {
+                contextRef.setupTest({
+                    phase: 'action',
+                    player1: {
+                        hand: ['jedi-lightsaber'],
+                        leader: { card: 'finn#this-is-a-rescue', deployed: false },
+                        groundArena: ['battlefield-marine'],
+                        resources: 5
+                    },
+                    player2: {
+                        hand: ['top-target'],
+                        groundArena: ['wampa'],
+                        resources: 5
+                    }
+                });
+            });
+
+            it('should defeat a friendly upgrade and give a shield token', function () {
+                const { context } = contextRef;
+
+                // Equip a friendly upgrade to battlefield marine
+                context.player1.clickCard(context.jediLightsaber);
+                expect(context.player1).toBeAbleToSelectExactly([context.battlefieldMarine, context.wampa]);
+                context.player1.clickCard(context.battlefieldMarine);
+                expect(context.battlefieldMarine).toHaveExactUpgradeNames(['jedi-lightsaber']);
+
+                context.player2.passAction();
+
+                // Use Finn's ability
+                context.player1.clickCard(context.finn);
+                context.player1.clickPrompt('Defeat a friendly upgrade on a unit. If you do, give a Shield token to that unit');
+                expect(context.battlefieldMarine).toHaveExactUpgradeNames(['shield']);
+            });
+
+            it('should not be able to defeat an opponent\'s (non-friendly) upgrade', function () {
+                const { context } = contextRef;
+
+                // Equip a friendly upgrade to battlefield marine
+                context.player1.clickCard(context.jediLightsaber);
+                expect(context.player1).toBeAbleToSelectExactly([context.battlefieldMarine, context.wampa]);
+                context.player1.clickCard(context.battlefieldMarine);
+                expect(context.battlefieldMarine).toHaveExactUpgradeNames(['jedi-lightsaber']);
+
+                // Opponent equips an upgrade to battlefield marine
+                context.player2.clickCard(context.topTarget);
+                context.player2.clickCard(context.battlefieldMarine);
+                expect(context.battlefieldMarine).toHaveExactUpgradeNames(['jedi-lightsaber', 'top-target']);
+
+                // Unable to use Finn's ability on non-friendly upgrade
+                context.player1.clickCard(context.finn);
+                context.player1.clickPrompt('Defeat a friendly upgrade on a unit. If you do, give a Shield token to that unit');
```

Comment:
Protip: you can set `resources: 4` for player1 so that there aren't enough resources to deploy and you won't get this prompt since his leader action is the only available on-click action